v2/report_publisher.go

```diff
@@ -1,14 +1,74 @@
 package bugsnag
 
-import "fmt"
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+)
 
 type reportPublisher interface {
 	publishReport(*payload) error
+	setMainProgramContext(context.Context)
 }
 
-type defaultReportPublisher struct{}
+func (defPub *defaultReportPublisher) delivery() {
+	signalsCh := make(chan os.Signal, 1)
+	signal.Notify(signalsCh, syscall.SIGINT, syscall.SIGTERM)
 
-func (*defaultReportPublisher) publishReport(p *payload) error {
+waitForEnd:
+	for {
+		select {
+		case <-signalsCh:
+			defPub.isClosing = true
+			break waitForEnd
+		case <-defPub.mainProgramCtx.Done():
+			defPub.isClosing = true
```

Comment:
Minor nit: This could be moved to right before `close(defPub.eventsChan)` since that's our common "this publisher is closing" path.